### PR TITLE
feat: add reduce to star routing

### PR DESCRIPTION
**Pull request by alaeddine-13**
closes: https://github.com/jina-ai/jina/issues/4021

### CODE REVIEW

File: jina/peapods/runtimes/head/__init__.py
Comment (jacobowitz) [inline jina/peapods/runtimes/head/__init__.py:250]: Lets move the other code here? I mean the reduce logic inside the elif. Maybe even inside the DataRequestHandler?
If it is before the if we may do unnecessary work (ok actually you are checking correctly in your if, but the place is still a bit surprising to me ;))

**Review by jacobowitz — CHANGES_REQUESTED** [review]
Now that I see the code I would say lets to the reducing inside the DataRequestHandler?

File: jina/peapods/runtimes/request_handlers/data_request_handler.py
Comment (jacobowitz) [inline jina/peapods/runtimes/request_handlers/data_request_handler.py:122]: I think there is no need to move this code up, lets leave it where it was before? at line 133?

File: jina/peapods/runtimes/head/__init__.py
Comment (jacobowitz) [inline jina/peapods/runtimes/head/__init__.py:215]: I would suggest to move this and the following line to the DataRequestHandler as a static function? Then there is less doc related logic here in this runtime and we dont have to copy those two lines again in line 244/245

**Review by jacobowitz — CHANGES_REQUESTED** [review]
just some minor style comments, looks great otherwise!

File: jina/peapods/runtimes/request_handlers/data_request_handler.py
Comment (alaeddine-13) [reply jina/peapods/runtimes/request_handlers/data_request_handler.py]: changed

File: jina/peapods/runtimes/head/__init__.py
Comment (alaeddine-13) [reply jina/peapods/runtimes/head/__init__.py]: I refactored this, can you re-check ?

**Review by jacobowitz — APPROVED** [review]
looks good :+1: 
Lets wait for CI and if its green we can merge

File: jina/types/request/data.py
Comment (JoanFM) [inline jina/types/request/data.py:25]: why we need this?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/request/data.py
Comment (alaeddine-13) [reply jina/types/request/data.py:25]: it's needed to make sure request objects and DocumentArrays extracted from the request point to the same protobuf object. This helps perform reduce in-place without having to copy back to the request object.
Actually we noticed that sometimes Documents extracted from the request object aren't views to the same protobuf so we wanted to enforce the pattern in all cases

Comment (JoanFM) [reply jina/types/request/data.py:25]: then this will need to be reviewed with the new docArray

**Review by jacobowitz — APPROVED** [review]
Looks good! I checked the failing tests and they are unrelated to your change. Actually they surfaced a bug in k8s when the connection pool is disabled. I will tackle this separately

**Comment by jacobowitz** [comment]
K8s follow up bug https://github.com/jina-ai/internal-tasks/issues/346

**Comment by alaeddine-13** [comment]
I see that tests/k8s/test_k8s.py::test_flow_with_needs is related to this PR
Actually the test fails because default reduce is applied right before custom merging.

**Pull request merged**
